### PR TITLE
Upstream sync a5267be64..6f8fe6242

### PR DESCRIPTION
**Pull request by RassK**
## What

Upstream sync a5267be6402a22f7404acb9d88b7ad3aef059094..6f8fe62424419c72621e33e39c2736207f156efd

~~**NB!** Disabling `SIGNALFX_HTTP_SERVER_TAG_QUERY_STRING` as this was enabled by default. We should look if we need it, document it and re-enable again.~~

**This feature is now actually public.**

### CODE REVIEW

**Review by aurbiztondo-splunk — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]
LGTM, minor left-overs (fine to come separate):

- Not from this change but there is a left-over with incorrect public key `def86d061d0d2eeb` at tracer\test\test-applications\aspnet\Samples.AspNet.VersionConflict\Samples.AspNet.VersionConflict.csproj
- Check GUID at profiler\test\Datadog.Profiler.IntegrationTests\Helpers\EnvironmentHelper.cs
- There a few env vars with DD_ prefix

File: tracer/build/_build/BuildVariables.cs [hunk 1]
```diff
@@ -40,8 +40,8 @@ public static void AddTracerEnvironmentVariables(this Dictionary<string, string>
 
         if (EnvironmentInfo.IsWin)
         {
-            envVars.Add("CORECLR_PROFILER_PATH_32", monitoringHomeDirectory / "win-x86" / "SignalFx.Tracing.ClrProfiler.Native.dll");
-            envVars.Add("CORECLR_PROFILER_PATH_64", monitoringHomeDirectory / "win-x64" / "SignalFx.Tracing.ClrProfiler.Native.dll");
+            envVars.Add("CORECLR_PROFILER_PATH_32", monitoringHomeDirectory / "Datadog.AutoInstrumentation.NativeLoader.x86.dll");
```
Comment (Kielek) [inline tracer/build/_build/BuildVariables.cs:43]: Please double check this.

File: tracer/build/_build/docker/centos7.dockerfile [hunk 2]
```diff
@@ -46,13 +46,15 @@ RUN yum update -y \
...
+# Install newer version of fpm and specific version of dotenv 
 RUN echo "gem: --no-document --no-rdoc --no-ri" > ~/.gemrc && \
     gem install --version 1.12.2 --user-install ffi && \
     gem install --version 1.6.0 --user-install git && \
     gem install --version 0.9.10 --user-install rb-inotify && \
     gem install --version 3.2.3  --user-install rexml && \
-    gem install backports -v 3.21.0 && gem install fpm
+    gem install backports -v 3.21.0 && \
+    gem install --version 2.7.6 dotenv && \
+    gem install fpm
```
Comment (Kielek) [inline tracer/build/_build/docker/centos7.dockerfile:57]: lack of --version for fpm. I see that it was done previously, you can create separate ticket for it.
Consider 1.14.1 like in debiean.dockerfile

File: tracer/src/Datadog.Trace/Configuration/ConfigurationKeys.cs [hunk 3]
```diff
@@ -405,6 +419,27 @@ internal static partial class ConfigurationKeys
         /// </summary>
         public const string TraceMethods = "SIGNALFX_TRACE_METHODS";
 
+        /// <summary>
+        /// Configuration key for specifying a custom regex to obfuscate query strings.
+        /// Default value is in TracerSettings
+        /// </summary>
+        /// <seealso cref="TracerSettings.ObfuscationQueryStringRegex"/>
+        public const string ObfuscationQueryStringRegex = "DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP";
```
Comment (Kielek) [inline tracer/src/Datadog.Trace/Configuration/ConfigurationKeys.cs:427]: DD_ -> SIGNALFX_
+ documentation

**Review by Kielek — APPROVED** [review]

**Review by dszmigielski — APPROVED** [review]

File: tracer/build/_build/BuildVariables.cs (hunk 1 above)
Comment (RassK) [reply tracer/build/_build/BuildVariables.cs:43]: It's bit older thing. Seems we missed the native loader renaming.

**Comment by RassK** [comment]
> * Not from this change but there is a left-over with incorrect public key `def86d061d0d2eeb` at tracer\test\test-applications\aspnet\Samples.AspNet.VersionConflict\Samples.AspNet.VersionConflict.csproj

It's actually coming from DD's Nuget package, which we also don't publish. I'm also not sure how accurate is the test now, that we try to auto solve version conflict between DD's tracer and our tracer version 🤔

**Review by Kielek — APPROVED** [review]

**Comment by RassK** [comment]
Updated changelog, moved new environment variables to `advanced-configuration.md`

**Review by Kielek — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]

File: docs/advanced-config.md [hunk 4]
```diff
@@ -158,3 +158,19 @@ of SignalFx Instrumentation for .NET.
...
 | `SIGNALFX_PROFILER_ENABLED` | Enable to activate thread sampling. | `false` |
 | `SIGNALFX_PROFILER_CALL_STACK_INTERVAL` | Sampling period. It defines how often the threads are stopped in order to fetch all stack traces. This value cannot be lower than `1000` milliseconds. | `10000` |
+
+## Including query string settings
+
+This feature is ASP.NET Core only.
+
+| Setting | Description | Default |
+|-|-|-|
+| `SIGNALFX_TRACE_OBFUSCATION_QUERY_STRING_REGEXP` | Specifying a custom regex to obfuscate query strings. | (see default below)[^regex] |
```
Comment (aurbiztondo-splunk) [inline docs/advanced-config.md:168]: I would use present tense in all rows (Specifies...)

File: docs/internal/internal-config.md [hunk 5]
```diff
@@ -75,12 +75,12 @@ These settings should be never used by the users.
...
 | `SIGNALFX_DEBUGGER_SNAPSHOT_URL` | Sets the URL used to query our backend directly for the list of active probes. This can only be used if SIGNALFX_API_KEY is also available. | |
 | `SIGNALFX_DEBUGGER_PROBE_FILE` | Sets the probe configuration file full path. Loads the probe configuration from a local file on disk. Useful for local development and testing. | |
-| `SIGNALFX_DEBUGGER_ENABLED` | Enabling or disabling the Live Debugger | `false`  |
 | `SIGNALFX_DEBUGGER_MAX_DEPTH_TO_SERIALIZE` | Sets the max object depth to serialize for probe snapshots. | `1`  |
 | `SIGNALFX_DEBUGGER_MAX_TIME_TO_SERIALIZE` | Sets the maximum duration (in milliseconds) to run serialization for probe snapshots. | `150`  |
 | `SIGNALFX_DEBUGGER_UPLOAD_BATCH_SIZE` | Sets the maximum upload batch size. | `100`  |
 | `SIGNALFX_DEBUGGER_DIAGNOSTICS_INTERVAL` | Sets the interval (in seconds) between sending probe statuses. | `3600`  |
 | `SIGNALFX_DEBUGGER_UPLOAD_FLUSH_INTERVAL` | Sets the interval (in milliseconds) between flushing statuses. | `0`  |
+| `SIGNALFX_INTERNAL_DEBUGGER_ENABLED` | Enabling or disabling the Live Debugger | `false`  |
```
Comment (aurbiztondo-splunk) [inline docs/internal/internal-config.md:83]: Same comment as above, use "Enables".

File: tracer/src/Datadog.Trace/Ci/CIVisibility.cs [hunk 6]
```diff
@@ -159,41 +159,47 @@ private static async Task ShutdownAsync()
 
         private static bool InternalEnabled()
         {
+            var processName = ProcessHelpers.GetCurrentProcessName() ?? string.Empty;
+
+            // By configuration
             if (_settings.Enabled)
             {
+                // When is enabled by configuration we only enable it to the testhost child process if the process name is dotnet.
```
Comment (aurbiztondo-splunk) [inline tracer/src/Datadog.Trace/Ci/CIVisibility.cs:167]: What do you mean here? I don't follow this sentence...

File: tracer/src/Datadog.Trace/ClrProfiler/Instrumentation.cs [hunk 7]
```diff
@@ -72,6 +72,14 @@ public static void Initialize()
                 return;
             }
 
+            if (CIVisibility.Settings.Enabled && !CIVisibility.Enabled)
+            {
+                // If CI Visibility is enabled by configuration
+                // we check if is the testhost.dll process
+                // we avoid instrumenting other process started from dotnet test.
+                return;
```
Comment (aurbiztondo-splunk) [inline tracer/src/Datadog.Trace/ClrProfiler/Instrumentation.cs:80]: Suggestion:
// If CI Visibility is enabled by configuration
// we check if it's the testhost.dll process
// we avoid instrumenting any other process started by the dotnet test.

File: tracer/src/Datadog.Trace/Configuration/ConfigurationKeys.cs [hunk 8]
```diff
@@ -340,6 +340,20 @@ internal static partial class ConfigurationKeys
         /// <seealso cref="TracerSettings.HttpClientErrorStatusCodes"/>
         public const string HttpClientErrorStatusCodes = "SIGNALFX_HTTP_CLIENT_ERROR_STATUSES";
 
+        /// <summary>
+        /// Configuration key indicating the optional name of the custom header to take into account to report the ip address from.
+        /// If this variable is set all other IP related headers should be ignored
+        /// Default is value is null (do not override).
```
Comment (aurbiztondo-splunk) [inline tracer/src/Datadog.Trace/Configuration/ConfigurationKeys.cs:346]: Default value is null (there's two is)

**Review by aurbiztondo-splunk — APPROVED** [review]
Left a couple of comments.

**Comment by Kielek** [comment]
@aurbiztondo-splunk, thanks for your feedback. Most of the file with `.cs` extensions are taken from the upstream repository. We are trying to avoid any modifications in such files.

The most important part is to keep `.md` files in good shape.

Thanks for your comments! :)

**Review by dszmigielski — APPROVED** [review]

File: tracer/src/Datadog.Trace/Ci/CIVisibility.cs (hunk 6 above)
Comment (RassK) [reply tracer/src/Datadog.Trace/Ci/CIVisibility.cs:167]: We can leave it for now, will cause conflicts with upstream.

File: tracer/src/Datadog.Trace/ClrProfiler/Instrumentation.cs (hunk 7 above)
Comment (RassK) [reply tracer/src/Datadog.Trace/ClrProfiler/Instrumentation.cs:80]: We can leave it for now, will cause conflicts with upstream.

File: tracer/src/Datadog.Trace/Configuration/ConfigurationKeys.cs (hunk 8 above)
Comment (RassK) [reply tracer/src/Datadog.Trace/Configuration/ConfigurationKeys.cs:346]: We can leave it for now, will cause conflicts with upstream.

**Comment by RassK** [comment]
@aurbiztondo-splunk thanks! updated. Feel free to use the Github "suggest changes" feature, so we can commit directly your fixes.

**Review by aurbiztondo-splunk — APPROVED** [review]

**Pull request merged**